### PR TITLE
Compact histograms on ingestion

### PR DESCRIPTION
**Pull request by beorn7**
_This is only for the sparsehistogram branch!_

This evolved into a bit of yak shaving.

Summary:
- Compact for (integer) Histogram was relatively easy thanks to generics. (I'm now considering getting rid of more duplicated code, but that's for another PR.) Note, however, that the delta buckets of integer Histograms had to be treated specially.
- While refining the doc comment for the Compact method, I realized that we probably shouldn't do the `maxEmptyBuckets==3` thing, so I removed it (by setting it to 0, not by removing the feature, it might still turn out useful).
- Finally I added the `Compact` call for both `Histogram` and `FloatHistogram` in the parser, which appeared to me to be the most reasonably place.
- Since I expect in most cases that the `Compact` call in the parser will be a no-op, I added a fast path to it for that case to minimize the performance impact.

### CODE REVIEW

**Comment by beorn7** [comment]
Hmmm, our golangci-lint doesn't know Go1.19 yet. That's why we cannot get all test pass.

**Comment by beorn7** [comment]
I'll just switch to Go1.18 then. Enough for generics.

**Comment by beorn7** [comment]
I realized there is a bug in the code. Will fix ASAP. Stay tuned…

**Comment by beorn7** [comment]
OK, it works now. Please review now. :o)

File: promql/functions.go [hunk 1]
```diff
@@ -31,18 +31,23 @@ import (
...
+// enh.Out is a pre-allocated empty vector that you may use to accumulate output
+// before returning it. The vectors in vals should not be returned.
+//
+// Range vector functions need only return a vector with the right value, the
+// metric and timestamp are not needed.
+//
 // Instant vector functions need only return a vector with the right values and
-//     metrics, the timestamp are not needed.
+// metrics, the timestamp are not needed.
+//
```
Comment (beorn7) [inline promql/functions.go:50]: Sorry for the reformatting here. It was weirdly formatted before this commit, and somehow Go1.18 didn't like the orginal formatting and formatted this even more weirdly. So I manually formatted it "normally".

File: promql/functions.go [hunk 2]
```diff
@@ -159,9 +164,7 @@ func extrapolatedRate(vals []parser.Value, args parser.Expressions, enh *EvalNod
 
 // histogramRate is a helper function for extrapolatedRate. It requires
 // points[0] to be a histogram. It returns nil if any other Point in points is
-// not a histogram. Currently, it also returns nil on mixed schemas or zero
-// thresholds in the histograms, because it cannot handle those schema changes
-// yet.
```
Comment (beorn7) [inline promql/functions.go:164]: Removed this because it _is_ actually handled properly for a while now.

File: model/histogram/histogram.go [hunk 3]
```diff
@@ -27,11 +27,11 @@ import (
...
-//   Bucket boundaries →              [-2,-1)  [-1,-0.5) [-0.5,-0.25) ... [-0.001,0.001] ... (0.25,0.5] (0.5,1]  (1,2] ....
-//                                       ↑        ↑           ↑                  ↑                ↑         ↑      ↑
-//   Zero bucket (width e.g. 0.001) →    |        |           |                  ZB               |         |      |
-//   Positive bucket indices →           |        |           |                          ...     -1         0      1    2    3
-//   Negative bucket indices →  3   2    1        0          -1       ...
+//	Bucket boundaries →              [-2,-1)  [-1,-0.5) [-0.5,-0.25) ... [-0.001,0.001] ... (0.25,0.5] (0.5,1]  (1,2] ....
+//	                                    ↑        ↑           ↑                  ↑                ↑         ↑      ↑
+//	Zero bucket (width e.g. 0.001) →    |        |           |                  ZB               |         |      |
+//	Positive bucket indices →           |        |           |                          ...     -1         0      1    2    3
+//	Negative bucket indices →  3   2    1        0          -1       ...
```
Comment (beorn7) [inline model/histogram/histogram.go:34]: Reformatting happened because of move to Go1.18.

File: model/histogram/float_histogram.go [hunk 4]
```diff
@@ -515,21 +385,21 @@ func compactBuckets(buckets []float64, spans []Span, maxEmptyBuckets int) ([]flo
...
+//     is partially within, partially outside of the new ZeroThreshold, can only
+//     happen together with a counter reset and therefore shortcuts to returning
+//     true.
 //
-// * Upon a decrease of the Schema, the buckets of the previous histogram are
-//   merged so that they match the new, lower-resolution schema (again without
-//   mutating the provided previous histogram).
+//   - Upon a decrease of the Schema, the buckets of the previous histogram are
+//     merged so that they match the new, lower-resolution schema (again without
+//     mutating the provided previous histogram).
```
Comment (beorn7) [inline model/histogram/float_histogram.go:402]: Formatting change done by Go1.18.

**Comment by beorn7** [comment]
Review hint: The actually important differences are just in the `compactBuckets` function. The diff isn't displaying that well. I recommend to copy out the old and new version of the function and diff them with a tool of your choice. Then you see all the special-case handling for delta buckets, and the rest is pretty much the same.

File: model/histogram/compact.go [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+				if deltaBuckets {
+					currentBucketAbsolute = 0
+				}
+				posInSpan += uint32(nEmpty)
+				continue
+			}
+			// In all other cases, we cut out the empty buckets.
+			if deltaBuckets && iBucket+nEmpty < len(buckets) {
+				currentBucketAbsolute = -buckets[iBucket]
+				buckets[iBucket+nEmpty] -= currentBucketAbsolute
```
Comment (codesome) [inline model/histogram/compact.go:125]: I propose this for ease of understanding
```suggestion
				buckets[iBucket+nEmpty] += buckets[iBucket]
```

File: model/histogram/histogram_test.go [hunk 6]
```diff
@@ -410,3 +410,259 @@ func TestHistogramToFloat(t *testing.T) {
...
+				NegativeSpans:   []Span{{-10, 2}},
+				NegativeBuckets: []int64{2, 3},
+			},
+		},
+	}
+
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			require.Equal(t, c.expected, c.in.Compact(c.maxEmptyBuckets))
+			// Has it also happened in-place?
```
Comment (codesome) [inline model/histogram/histogram_test.go:664]: I feel it is easier to read test when we clearly state what needs to happen (else I have to read it as question and try to answer myself by looking at the code)
```suggestion
			// Compact also happens in-place.
```

**Review by codesome — APPROVED** [review]
LGTM with readability nits

File: model/histogram/histogram_test.go (hunk 6 above)
Comment (beorn7) [reply model/histogram/histogram_test.go:664]: Sure. I also changed in in float_histogram_test.go.

**Pull request merged**